Adds optional Extension property to set extension on uploaded artifact

georgealton opened the pull request:
*Issue #, if available:* #643 

*Description of changes:*
Some AWS services require that the objects they access in S3 have a particular extension.

This meant rain was previously incompatbile with that requirement.

this change allows a user to specify the extension to append to the uploaded artifact in S3.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

georgealton commented on README.md:
````diff
@@ -269,6 +269,28 @@ Resources:
         S3Key: 1b4844dacc843f09941c11c94f80981d3be8ae7578952c71e875ef7add37b1a7
 ```
 
+Sometimes you require uploaded objects to have a specific extension
````
```suggestion
Sometimes you require objects upload to S3 have a specific extension, use the `Extension` property to ensure the the artifact in S3 ends .<Extension>.
```

ericzbeard reviewed (changes requested):
Docs, including README.md, are auto-generated using `go generate ./...`

I'm also wondering about the key written into the template during packaging of assets. This change modifies the key in S3, but does it also handle the key in the template?

georgealton commented:
Oops :flushed: sorry about, thanks, moved from README.md -> README.tmpl

It does modify the template - but there's no test to assert this. cft/pkg/util.go takes the key returned by s3.Upload 

I've added to scripts/integ.sh to run rain package, but not asserting on the output. I Could pipe the rain pkg output in yq or just grep to verify that the value ends with the Extension...
